README
libsvg-cairo - Render SVG content using cairo

Description
-----------
libsvg-cairo provides the ability to render SVG content from files or
buffers. All rendering is performed using the cairo rendering library.

Dependencies
------------
libsvg-cairo depends on libsvg and cairo. It should be easy to build on
any reasonable system.

Documentation
-------------
libsvg-cairo is accessed by including its public header file:

#include <svg-cairo.h>

For now, that header file contains all existing documentation. But it
was intentionally designed to be very intuitive to use without
excessive explanations.

History
-------
libsvg-cairo was developed by Carl Worth <cworth@isi.edu>


NEWS
Snapshot 0.4.0 (2009-03-23 Philip de Nier <philipn@users.sourceforge.net>)
==========================================================================
- Various changes (some as placeholders) to match changes to libsvg (v0.3-0.4)

Bug fixes
---------
- fix percentage lengths which should be relative to the view box width or 
height


Snapshot 0.3.0 (2009-02-28 Philip de Nier <philipn@users.sourceforge.net>)
==========================================================================
- Various changes (some as placeholders) to match changes to libsvg (v0.2-0.3)


Snapshot 0.2.0 (2009-02-10 Philip de Nier <philipn@users.sourceforge.net>)
==========================================================================
- Copied code from cairo graphics CVS
- Add support for compiling on Windows
- Added functions to allow a different svg_t parse tree to be rendered from and
a function to return the render engine
- Changes to match changes to the libsvg API 0.2.0

Bug fixes
---------
- type casts etc. to avoid MSVC compiler warnings


Snapshot 0.1.6 (2005-06-15 Carl Worth <cworth@cworth.org>)
==========================================================
Track all changes in cairo API up to cairo 0.5.0 (which is now the 
minimum version of cairo required).

Bug fixes
---------
- Fix autogen.sh to call all commands through environment variables.

Snapshot 0.1.5 (2005-02-13 Carl Worth <cworth@cworth.org>)
==========================================================
Big performance fix for SVG images with deep grouping within a group
with opacity != 1.0, (thanks to TOKUNAGA Hiroyuki)

Various bug fixes.

Older snapshots
===============

2003-11-15  Richard Worth <richard@theworths.org>

        * Renamed libxsvg to libsvg-cairo

2002-??-??  Carl Worth  <cworth@east.isi.edu>

        * Initial version of libxsvg


win32_config.h
/* win32_config.h  Manually edited from config.h */
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.in by autoheader.  */

/* Define to 1 if you have the <dlfcn.h> header file. */
/* #undef HAVE_DLFCN_H */

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Has pangocairo module */
/* #undef HAVE_PANGOCAIRO */

/* Define to 1 if you have the <stdint.h> header file. */
/* #undef HAVE_STDINT_H */

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
/* #undef HAVE_STRINGS_H */

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
/* #undef HAVE_UNISTD_H */

/* Name of package */
#define PACKAGE "libsvg-cairo"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME ""

/* Define to the full name and version of this package. */
#define PACKAGE_STRING ""

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME ""

/* Define to the version of this package. */
#define PACKAGE_VERSION ""

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "0.1.6-head"

/* Define to 1 if the X Window System is missing or not being used. */
#define X_DISPLAY_MISSING 1


src/svg-cairo.h
/* libsvg-cairo - Render SVG documents using the cairo library
 *
 * Copyright © 2002 University of Southern California
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Author: Carl D. Worth <cworth@isi.edu>
 */

#ifndef SVG_CAIRO_H
#define SVG_CAIRO_H

#include <cairo.h>

#include <svg.h>

#ifdef __cplusplus
extern "C" {
#endif

typedef enum svg_cairo_status {
    SVG_CAIRO_STATUS_SUCCESS = SVG_STATUS_SUCCESS,
    SVG_CAIRO_STATUS_NO_MEMORY = SVG_STATUS_NO_MEMORY,
    SVG_CAIRO_STATUS_IO_ERROR = SVG_STATUS_IO_ERROR,
    SVG_CAIRO_STATUS_FILE_NOT_FOUND = SVG_STATUS_FILE_NOT_FOUND,
    SVG_CAIRO_STATUS_INVALID_VALUE = SVG_STATUS_INVALID_VALUE,
    SVG_CAIRO_STATUS_INVALID_CALL = SVG_STATUS_INVALID_CALL,
    SVG_CAIRO_STATUS_PARSE_ERROR = SVG_STATUS_PARSE_ERROR
} svg_cairo_status_t;

typedef struct svg_cairo svg_cairo_t;

svg_cairo_status_t
svg_cairo_create (svg_cairo_t **svg_cairo);

svg_cairo_status_t
svg_cairo_destroy (svg_cairo_t *svg_cairo);

svg_cairo_status_t
svg_cairo_parse (svg_cairo_t *svg_cairo, const char *uri_or_filename);

svg_status_t
svg_cairo_set_base_directory (svg_cairo_t *svg_cairo, const char *directory);

svg_status_t
svg_cairo_set_base_uri (svg_cairo_t *svg_cairo, const char *abs_uri);

svg_cairo_status_t
svg_cairo_parse_file (svg_cairo_t *svg_cairo, FILE *file);

svg_cairo_status_t
svg_cairo_parse_buffer (svg_cairo_t *svg_cairo, const char *buf, size_t count);

svg_cairo_status_t
svg_cairo_parse_chunk_begin (svg_cairo_t *svg_cairo);

svg_cairo_status_t
svg_cairo_parse_chunk       (svg_cairo_t *svg_cairo, const char *buf, size_t count);

svg_cairo_status_t
svg_cairo_parse_chunk_end   (svg_cairo_t *svg_cairo);

svg_cairo_status_t
svg_cairo_render (svg_cairo_t *svg_cairo, cairo_t *xrs);

/* XXX philipn: intended to replace svg_cairo_render */
svg_cairo_status_t
svg_cairo_render_2 (svg_cairo_t *svg_cairo, svg_t *svg, cairo_t *xrs);

svg_cairo_status_t
svg_cairo_render_element (svg_cairo_t 	*svg_cairo, 
			  svg_t		*svg, 
			  svg_element_t	*element,
			  cairo_t 	*xrs);

/* XXX philipn: generalize to preparing the engine before svg_cairo_get_render_engine */
svg_cairo_status_t
svg_cairo_set_drawing_context (svg_cairo_t *svg_cairo, cairo_t *xrs);

/* XXX: Ugh... this inconsistent interface needs to be cleaned up. */
svg_cairo_status_t
svg_cairo_set_viewport_dimension (svg_cairo_t *svg_cairo, unsigned int width, unsigned int height);

void
svg_cairo_get_size (svg_cairo_t  *svg_cairo,
		    unsigned int *width,
		    unsigned int *height);

void
svg_cairo_get_render_engine (svg_cairo_t *svg_cairo, svg_render_engine_t **engine, void **closure);

#ifdef __cplusplus
}
#endif

#endif


src/svg_cairo_state.c
/* libsvg-cairo - Render SVG documents using the cairo library
 *
 * Copyright © 2002 University of Southern California
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Author: Carl D. Worth <cworth@isi.edu>
 */

#include <stdlib.h>
#include <string.h>

#include "svg-cairo-internal.h"

svg_cairo_status_t
_svg_cairo_state_create (svg_cairo_state_t **state)
{
    *state = malloc (sizeof (svg_cairo_state_t));
    if (*state == NULL)
	return SVG_CAIRO_STATUS_NO_MEMORY;

    _svg_cairo_state_init (*state);

    return SVG_CAIRO_STATUS_SUCCESS;
}

svg_cairo_status_t
_svg_cairo_state_init (svg_cairo_state_t *state)
{
    /* trust libsvg to set all of these to reasonable defaults:
    state->fill_paint;
    state->stroke_paint;
    state->fill_opacity;
    state->stroke_opacity;
    */
    state->child_surface = NULL;
    state->saved_cr = NULL;

#if HAVE_PANGOCAIRO
    state->font_description = pango_font_description_new ();
#else
    state->font_family = strdup (SVG_CAIRO_FONT_FAMILY_DEFAULT);
    if (state->font_family == NULL)
	return SVG_CAIRO_STATUS_NO_MEMORY;

    state->font_size = 1.0;
    state->font_style = SVG_FONT_STYLE_NORMAL;
    state->font_weight = 400;
    state->font_dirty = 1;
#endif

    state->dash = NULL;
    state->num_dashes = 0;
    state->dash_offset = 0;

    state->opacity = 1.0;

    state->bbox = 0;

    state->text_anchor = SVG_TEXT_ANCHOR_START;

    state->next = NULL;

    return SVG_CAIRO_STATUS_SUCCESS;
}

svg_cairo_status_t
_svg_cairo_state_init_copy (svg_cairo_state_t *state, const svg_cairo_state_t *other)
{
    _svg_cairo_state_deinit (state);

    if (other == NULL)
	return _svg_cairo_state_init (state);

    *state = *other;

    /* We don't need our own child_surface or saved cr at this point. */
    state->child_surface = NULL;
    state->saved_cr = NULL;

#if HAVE_PANGOCAIRO
    state->font_description = pango_font_description_copy (other->font_description);
#else
    if (other->font_family)
	state->font_family = strdup ((char *) other->font_family);
#endif

    state->viewport_width = other->viewport_width;
    state->viewport_height = other->viewport_height;
    state->view_box_width = other->view_box_width;
    state->view_box_height = other->view_box_height;

    if (other->dash) {
	state->dash = malloc (state->num_dashes * sizeof(double));
	if (state->dash == NULL)
	    return SVG_CAIRO_STATUS_NO_MEMORY;
	memcpy (state->dash, other->dash, state->num_dashes * sizeof(double));
    }

    return SVG_CAIRO_STATUS_SUCCESS;
}

svg_cairo_status_t
_svg_cairo_state_deinit (svg_cairo_state_t *state)
{
    if (state->child_surface) {
	cairo_surface_destroy(state->child_surface);
	state->child_surface = NULL;
    }

    if (state->saved_cr) {
	cairo_destroy(state->saved_cr);
	state->saved_cr = NULL;
    }

#if HAVE_PANGOCAIRO
    if (state->font_description) {
	pango_font_description_free (state->font_description);
	state->font_description = NULL;
    }
#else
    if (state->font_family) {
	free (state->font_family);
	state->font_family = NULL;
    }
#endif

    if (state->dash) {
	free (state->dash);
	state->dash = NULL;
    }

    state->next = NULL;

    return SVG_CAIRO_STATUS_SUCCESS;
}

svg_cairo_status_t
_svg_cairo_state_destroy (svg_cairo_state_t *state)
{
    _svg_cairo_state_deinit (state);

    free (state);

    return SVG_CAIRO_STATUS_SUCCESS;
}

svg_cairo_state_t *
_svg_cairo_state_push (svg_cairo_state_t *state)
{
    svg_cairo_state_t *new;

    _svg_cairo_state_create (&new);
    if (new == NULL)
	return NULL;

    _svg_cairo_state_init_copy (new, state);

    new->next = state;

    return new;
}

svg_cairo_state_t *
_svg_cairo_state_pop (svg_cairo_state_t *state)
{
    svg_cairo_state_t *next;

    if (state == NULL)
	return NULL;

    next = state->next;

    _svg_cairo_state_destroy (state);

    return next;
}


src/svg_cairo_sprintf_alloc.c
/* libsvg-cairo - Render SVG documents using the cairo library
 *
 * Copyright © 2002 University of Southern California
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Author: Carl D. Worth <cworth@isi.edu>
 */

#include <stdlib.h>
#include <stdarg.h>
#define __USE_ISOC99
#include <stdio.h>

#include "svg-cairo-internal.h"

int
_svg_cairo_sprintf_alloc (char **str, const char *fmt, ...)
{
    int ret;
    va_list ap;

    va_start(ap, fmt);
    ret = _svg_cairo_vsprintf_alloc(str, fmt, ap);
    va_end(ap);

    return ret;
}

/* ripped more or less straight out of PRINTF(3) */
int
_svg_cairo_vsprintf_alloc (char **str, const char *fmt, va_list ap)
{
    char *new_str;
    /* Guess we need no more than 100 bytes. */
    int n, size = 100;
 
    if ((*str = malloc (size)) == NULL)
	return -1;
    while (1) {
	/* Try to print in the allocated space. */
	n = vsnprintf (*str, size, fmt, ap);
	/* If that worked, return the size. */
	if (n > -1 && n < size)
	    return n;
	/* Else try again with more space. */
	if (n > -1)    /* glibc 2.1 */
	    size = n+1; /* precisely what is needed */
	else           /* glibc 2.0 */
	    size *= 2;  /* twice the old size */
	new_str = realloc(*str, size);
	if (new_str == NULL) {
	    free(*str);
	    *str = NULL;
	    return -1;
	}
	*str = new_str;
    }
}


src/svg-cairo-internal.h
/* libsvg-cairo - Render SVG documents using the cairo library
 *
 * Copyright © 2002 University of Southern California
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Author: Carl D. Worth <cworth@isi.edu>
 */

#ifndef SVG_CAIRO_INTERNAL_H
#define SVG_CAIRO_INTERNAL_H
#if HAVE_CONFIG_H
#if !defined(_MSC_VER)
#include <config.h>
#endif
#endif

#include "svg-cairo.h"
#if !defined(_MSC_VER)
/* XXX: defines in svg-cairo-version.h are currently not used, but what can be done
to include svg-cairo-version.h.in in Windows builds */
#include "svg-cairo-version.h"
#endif

#include <stdarg.h>

#if HAVE_PANGOCAIRO
#include <pango/pango.h>
#include <pango/pangocairo.h>
#endif

/* XXX: What should this actually be? */
#define SVG_CAIRO_FONT_FAMILY_DEFAULT "sans-serif"

typedef struct svg_cairo_pt {
    double x;
    double y;
} svg_cairo_pt_t;

typedef enum svg_cairo_render_type {
    SVG_CAIRO_RENDER_TYPE_FILL,
    SVG_CAIRO_RENDER_TYPE_STROKE
} svg_cairo_render_type_t;

typedef struct svg_cairo_state {
    cairo_surface_t *child_surface;
    cairo_t *saved_cr;

    svg_color_t color;

    svg_paint_t fill_paint;
    svg_paint_t stroke_paint;
    double fill_opacity;
    double stroke_opacity;

#if HAVE_PANGOCAIRO
    PangoFontDescription *font_description;
#else
    char *font_family;
    double font_size;
    svg_font_style_t font_style;
    unsigned int font_weight;
    int font_dirty;
#endif

    double *dash;
    int num_dashes;
    double dash_offset;

    double opacity;

    double viewport_width;
    double viewport_height;
    double view_box_width;
    double view_box_height;

    int bbox;

    svg_text_anchor_t text_anchor;

    struct svg_cairo_state *next;
} svg_cairo_state_t;

struct svg_cairo {
    svg_t *svg;
    cairo_t *cr;

    svg_cairo_state_t *state;

    unsigned int viewport_width;
    unsigned int viewport_height;
    
    double text_chunk_width;
};

/* svg_cairo_sprintf_alloc.c */

int
_svg_cairo_sprintf_alloc (char **str, const char *fmt, ...);

int
_svg_cairo_vsprintf_alloc (char **str, const char *fmt, va_list ap);

/* svg_cairo_state.c */

svg_cairo_status_t
_svg_cairo_state_create (svg_cairo_state_t **state);

svg_cairo_status_t
_svg_cairo_state_init (svg_cairo_state_t *state);

svg_cairo_status_t
_svg_cairo_state_init_copy (svg_cairo_state_t *state, const svg_cairo_state_t *other);

svg_cairo_status_t
_svg_cairo_state_deinit (svg_cairo_state_t *state);

svg_cairo_status_t
_svg_cairo_state_destroy (svg_cairo_state_t *state);

svg_cairo_state_t *
_svg_cairo_state_push (svg_cairo_state_t *state);

svg_cairo_state_t *
_svg_cairo_state_pop (svg_cairo_state_t *state);


#endif
